File: apps/api/src/services/jobParsing.ts
/**
 * Pure rule-based extraction (regex/keyword scan) over raw job posting
 * text — no LLM call. A field that can't be confidently found returns an
 * honest default ("Not specified" / empty list), never a fabricated
 * stand-in value. Ported from HireConnect's job_parsing_logic.py.
 */
import {
  ATS_KEYWORD_VOCAB,
  DEFAULT_EXPERIENCE_LEVEL,
  DEFAULT_LOCATION,
  EXPERIENCE_LEVEL_PATTERNS,
  HIRING_TEAM_NAME_PATTERNS,
  KNOWN_SKILL_KEYWORDS,
  LOCATION_PATTERNS,
} from "../config/jobParsingVocab.js";

export interface ParsedJobPosting {
  roleTitle: string;
  company: string;
  skills: string[];
  experienceLevel: string;
  location: string;
  hiringTeamNames: string[];
  atsKeywords: string[];
  parsingConfidence: "high" | "low";
}

// Company group is non-greedy, stops at a sentence boundary — without that lookahead it
// swallows past "CloudTech." into the next sentence.
const ROLE_AT_COMPANY_RE = /^\s*([^.\n]{3,80}?)\s+at\s+([A-Z][\w&' -]{1,60}?)(?=[.,\n]|$)/;

// 2-8 consecutive ALL-CAPS words — many career sites render the title in caps.
const ALL_CAPS_PHRASE_RE = /\b([A-Z][A-Z]+(?:\s+[A-Z][A-Z]+){1,7})\b/;

const WORKDAY_HOST_RE = /^([a-z0-9-]+)\.wd\d+\.myworkdayjobs\.com$/i;
const GREENHOUSE_HOST_RE = /^(?:job-)?boards\.greenhouse\.io$/i;
const LEVER_HOST_RE = /^jobs\.lever\.co$/i;

const JOB_BOARD_HOSTS = new Set([
  "linkedin.com", "www.linkedin.com",
  "indeed.com", "www.indeed.com",
  "glassdoor.com", "www.glassdoor.com",
  "ziprecruiter.com", "www.ziprecruiter.com",
  "monster.com", "www.monster.com",
]);
const GENERIC_SUBDOMAIN_PREFIXES = new Set([
  "www", "careers", "career", "jobs", "job", "apply", "recruiting", "recruitment",
  "search", "search-jobs", "talent", "hiring", "workwith", "join", "emea", "us", "uk",
]);

function slugToName(slug: string): string {
  return slug
    .split(/[-_]+/)
    .filter(Boolean)
    .map((w) => w[0]!.toUpperCase() + w.slice(1))
    .join(" ");
}

function isTldLike(candidate: string, labelsLeft: number): boolean {
  return labelsLeft === 1 && (candidate.length <= 3 || ["com", "net", "org", "info"].includes(candidate));
}

/** Several major ATS platforms encode the employer directly in the URL — Workday's tenant
 * subdomain, Greenhouse/Lever's first path segment. Job boards are explicitly excluded. */
function companyFromUrl(jobUrl: string): string | null {
  let parsed: URL;
  try {
    parsed = new URL(jobUrl);
  } catch {
    return null;
  }
  const host = parsed.hostname.toLowerCase();
  if (JOB_BOARD_HOSTS.has(host)) return null;

  const workdayMatch = WORKDAY_HOST_RE.exec(host);
  if (workdayMatch) return slugToName(workdayMatch[1]!);

  const pathSegments = parsed.pathname.split("/").filter(Boolean);
  if ((GREENHOUSE_HOST_RE.test(host) || LEVER_HOST_RE.test(host)) && pathSegments.length > 0) {
    return slugToName(pathSegments[0]!);
  }

  // Strips generic prefixes repeatedly — real hosts stack them ("search.jobs.barclays").
  let labels = host.split(".").filter(Boolean);
  while (labels.length > 1 && GENERIC_SUBDOMAIN_PREFIXES.has(labels[0]!)) {
    labels = labels.slice(1);
  }
  if (labels.length > 0) {
    const candidate = labels[0]!;
    if (!GENERIC_SUBDOMAIN_PREFIXES.has(candidate) && !isTldLike(candidate, labels.length)) {
      return slugToName(candidate);
    }
  }
  return null;
}

function extractRoleAndCompany(rawText: string): { roleTitle: string | null; company: string | null } {
  const match = ROLE_AT_COMPANY_RE.exec(rawText);
  if (!match) return { roleTitle: null, company: null };
  const roleTitle = match[1]!.trim();
  const company = match[2]!.trim().replace(/[.,]+$/, "");
  return { roleTitle: roleTitle || null, company: company || null };
}

const TITLE_SEPARATOR_RE = /\s+[|•·–—]\s+|\s+-\s+|\s+@\s+/;
const COMPANY_TITLE_NOISE_RE = /\b(careers?|jobs?|job\s+search|hiring|vacancies|openings|apply|home)\b/gi;

function cleanCompanyFragment(fragment: string): string | null {
  const cleaned = fragment
    .replace(COMPANY_TITLE_NOISE_RE, "")
    .replace(/^[\s|\-–—·•,.]+|[\s|\-–—·•,.]+$/g, "")
    .replace(/\s{2,}/g, " ");
  return cleaned || null;
}

/** Splits the role off a page <title> like "Data Engineer | Barclays" — only when the
 * right-hand fragment is corroborated by the company the URL already independently yielded. */
function roleFromTitleLine(rawText: string, urlCompany: string | null): string | null {
  if (!urlCompany) return null;
  const lines = rawText.split("\n").map((l) => l.trim()).filter(Boolean);
  if (lines.length === 0) return null;

  const title = lines[0]!;
  if (title.length > 160) return null;

  const parts = title.split(TITLE_SEPARATOR_RE).map((p) => p.trim()).filter(Boolean);
  if (parts.length < 2) return null;

  const tail = cleanCompanyFragment(parts[parts.length - 1]!);
  if (!tail || tail.toLowerCase() !== urlCompany.toLowerCase()) return null;

  const role = parts[0]!.trim();
  COMPANY_TITLE_NOISE_RE.lastIndex = 0;
  if (new RegExp(`^(?:${COMPANY_TITLE_NOISE_RE.source})$`, "i").test(role)) return null;
  return role.length >= 3 && role.length <= 100 ? role : null;
}

function fallbackRoleTitle(rawText: string): string | null {
  const lines = rawText.split("\n").map((l) => l.trim()).filter(Boolean);
  if (lines.length < 2) return null;
  const first = lines[0]!;
  return first.length >= 3 && first.length <= 100 ? first : null;
}

function fallbackRoleTitleFromCaps(rawText: string): string | null {
  const match = ALL_CAPS_PHRASE_RE.exec(rawText);
  if (!match) return null;
  const phrase = match[1]!.trim();
  if (phrase.length < 6 || phrase.length > 80) return null;
  return phrase.replace(/\w\S*/g, (w) => w[0]!.toUpperCase() + w.slice(1).toLowerCase());
}

function matchVocabulary(rawText: string, vocabulary: string[]): string[] {
  const lowered = rawText.toLowerCase();
  return vocabulary.filter((term) => lowered.includes(term.toLowerCase()));
}

function matchFirstPattern(rawText: string, patterns: [RegExp, string][], fallback: string): string {
  for (const [pattern, label] of patterns) {
    if (pattern.test(rawText)) return label;
  }
  return fallback;
}

function extractHiringTeamNames(rawText: string): string[] {
  const names: string[] = [];
  for (const pattern of HIRING_TEAM_NAME_PATTERNS) {
    pattern.lastIndex = 0;
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(rawText))) {
      const name = match[1]!.trim();
      if (name && !names.includes(name)) names.push(name);
    }
  }
  return names;
}

export function parseJobPosting(rawText: string, jobUrl = ""): ParsedJobPosting {
  const text = rawText || "";
  const urlCompany = jobUrl ? companyFromUrl(jobUrl) : null;

  let { roleTitle, company } = extractRoleAndCompany(text);
  if (!roleTitle) roleTitle = roleFromTitleLine(text, urlCompany);
  if (!roleTitle) roleTitle = fallbackRoleTitle(text);
  if (!roleTitle) roleTitle = fallbackRoleTitleFromCaps(text);
  if (!company) company = urlCompany;

  const skills = matchVocabulary(text, KNOWN_SKILL_KEYWORDS);
  const atsKeywords = matchVocabulary(text, ATS_KEYWORD_VOCAB);
  const experienceLevel = matchFirstPattern(text, EXPERIENCE_LEVEL_PATTERNS, DEFAULT_EXPERIENCE_LEVEL);
  const location = matchFirstPattern(text, LOCATION_PATTERNS, DEFAULT_LOCATION);
  const hiringTeamNames = extractHiringTeamNames(text);

  return {
    roleTitle: roleTitle || "Not specified",
    company: company || "Not specified",
    skills,
    experienceLevel,
    location,
    hiringTeamNames,
    atsKeywords,
    parsingConfidence: roleTitle && company ? "high" : "low",
  };
}

File: apps/backend/services/job_parsing.py
"""Pure rule-based extraction (regex/keyword scan) over raw job posting
text — no LLM call. A field that can't be confidently found returns an
honest default ("Not specified" / empty list), never a fabricated
stand-in value. Mirrors apps/api's services/jobParsing.ts exactly.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal
from urllib.parse import urlparse

from config.job_parsing_vocab import (
    ATS_KEYWORD_VOCAB,
    DEFAULT_EXPERIENCE_LEVEL,
    DEFAULT_LOCATION,
    EXPERIENCE_LEVEL_PATTERNS,
    HIRING_TEAM_NAME_PATTERNS,
    KNOWN_SKILL_KEYWORDS,
    LOCATION_PATTERNS,
)


@dataclass
class ParsedJobPosting:
    roleTitle: str
    company: str
    skills: list[str]
    experienceLevel: str
    location: str
    hiringTeamNames: list[str]
    atsKeywords: list[str]
    parsingConfidence: Literal["high", "low"]


# Company group is non-greedy, stops at a sentence boundary — without that lookahead it
# swallows past "CloudTech." into the next sentence.
_ROLE_AT_COMPANY_RE = re.compile(r"^\s*([^.\n]{3,80}?)\s+at\s+([A-Z][\w&' -]{1,60}?)(?=[.,\n]|$)")

# 2-8 consecutive ALL-CAPS words — many career sites render the title in caps.
_ALL_CAPS_PHRASE_RE = re.compile(r"\b([A-Z][A-Z]+(?:\s+[A-Z][A-Z]+){1,7})\b")

_WORKDAY_HOST_RE = re.compile(r"^([a-z0-9-]+)\.wd\d+\.myworkdayjobs\.com$", re.IGNORECASE)
_GREENHOUSE_HOST_RE = re.compile(r"^(?:job-)?boards\.greenhouse\.io$", re.IGNORECASE)
_LEVER_HOST_RE = re.compile(r"^jobs\.lever\.co$", re.IGNORECASE)

_JOB_BOARD_HOSTS = {
    "linkedin.com", "www.linkedin.com",
    "indeed.com", "www.indeed.com",
    "glassdoor.com", "www.glassdoor.com",
    "ziprecruiter.com", "www.ziprecruiter.com",
    "monster.com", "www.monster.com",
}  # fmt: skip
_GENERIC_SUBDOMAIN_PREFIXES = {
    "www", "careers", "career", "jobs", "job", "apply", "recruiting", "recruitment",
    "search", "search-jobs", "talent", "hiring", "workwith", "join", "emea", "us", "uk",
}  # fmt: skip

_TITLE_SEPARATOR_RE = re.compile(r"\s+[|•·–—]\s+|\s+-\s+|\s+@\s+")
_COMPANY_TITLE_NOISE_RE = re.compile(r"\b(careers?|jobs?|job\s+search|hiring|vacancies|openings|apply|home)\b", re.IGNORECASE)


def _slug_to_name(slug: str) -> str:
    return " ".join(w[0].upper() + w[1:] for w in re.split(r"[-_]+", slug) if w)


def _is_tld_like(candidate: str, labels_left: int) -> bool:
    return labels_left == 1 and (len(candidate) <= 3 or candidate in ("com", "net", "org", "info"))


def _company_from_url(job_url: str) -> str | None:
    """Several major ATS platforms encode the employer directly in the URL — Workday's tenant
    subdomain, Greenhouse/Lever's first path segment. Job boards are explicitly excluded."""
    try:
        parsed = urlparse(job_url)
        if not parsed.hostname:
            return None
    except ValueError:
        return None

    host = parsed.hostname.lower()
    if host in _JOB_BOARD_HOSTS:
        return None

    workday_match = _WORKDAY_HOST_RE.match(host)
    if workday_match:
        return _slug_to_name(workday_match.group(1))

    path_segments = [s for s in parsed.path.split("/") if s]
    if (_GREENHOUSE_HOST_RE.match(host) or _LEVER_HOST_RE.match(host)) and path_segments:
        return _slug_to_name(path_segments[0])

    # Strips generic prefixes repeatedly — real hosts stack them ("search.jobs.barclays").
    labels = [label for label in host.split(".") if label]
    while len(labels) > 1 and labels[0] in _GENERIC_SUBDOMAIN_PREFIXES:
        labels = labels[1:]
    if labels:
        candidate = labels[0]
        if candidate not in _GENERIC_SUBDOMAIN_PREFIXES and not _is_tld_like(candidate, len(labels)):
            return _slug_to_name(candidate)
    return None


def _extract_role_and_company(raw_text: str) -> tuple[str | None, str | None]:
    match = _ROLE_AT_COMPANY_RE.match(raw_text)
    if not match:
        return None, None
    role_title = match.group(1).strip()
    company = re.sub(r"[.,]+$", "", match.group(2).strip())
    return role_title or None, company or None


def _clean_company_fragment(fragment: str) -> str | None:
    cleaned = _COMPANY_TITLE_NOISE_RE.sub("", fragment)
    cleaned = re.sub(r"^[\s|\-–—·•,.]+|[\s|\-–—·•,.]+$", "", cleaned)
    cleaned = re.sub(r"\s{2,}", " ", cleaned)
    return cleaned or None


def _role_from_title_line(raw_text: str, url_company: str | None) -> str | None:
    """Splits the role off a page <title> like "Data Engineer | Barclays" — only when the
    right-hand fragment is corroborated by the company the URL already independently yielded."""
    if not url_company:
        return None
    lines = [line.strip() for line in raw_text.split("\n") if line.strip()]
    if not lines:
        return None

    title = lines[0]
    if len(title) > 160:
        return None

    parts = [p.strip() for p in _TITLE_SEPARATOR_RE.split(title) if p.strip()]
    if len(parts) < 2:
        return None

    tail = _clean_company_fragment(parts[-1])
    if not tail or tail.lower() != url_company.lower():
        return None

    role = parts[0].strip()
    if re.fullmatch(_COMPANY_TITLE_NOISE_RE.pattern, role, re.IGNORECASE):
        return None
    return role if 3 <= len(role) <= 100 else None


def _fallback_role_title(raw_text: str) -> str | None:
    lines = [line.strip() for line in raw_text.split("\n") if line.strip()]
    if len(lines) < 2:
        return None
    first = lines[0]
    return first if 3 <= len(first) <= 100 else None


def _fallback_role_title_from_caps(raw_text: str) -> str | None:
    match = _ALL_CAPS_PHRASE_RE.search(raw_text)
    if not match:
        return None
    phrase = match.group(1).strip()
    if len(phrase) < 6 or len(phrase) > 80:
        return None
    return re.sub(r"\w\S*", lambda w: w.group(0)[0].upper() + w.group(0)[1:].lower(), phrase)


def _match_vocabulary(raw_text: str, vocabulary: list[str]) -> list[str]:
    lowered = raw_text.lower()
    return [term for term in vocabulary if term.lower() in lowered]


def _match_first_pattern(raw_text: str, patterns: list[tuple[re.Pattern[str], str]], fallback: str) -> str:
    for pattern, label in patterns:
        if pattern.search(raw_text):
            return label
    return fallback


def _extract_hiring_team_names(raw_text: str) -> list[str]:
    names: list[str] = []
    for pattern in HIRING_TEAM_NAME_PATTERNS:
        for match in pattern.finditer(raw_text):
            name = match.group(1).strip()
            if name and name not in names:
                names.append(name)
    return names


def parse_job_posting(raw_text: str, job_url: str = "") -> ParsedJobPosting:
    text = raw_text or ""
    url_company = _company_from_url(job_url) if job_url else None

    role_title, company = _extract_role_and_company(text)
    if not role_title:
        role_title = _role_from_title_line(text, url_company)
    if not role_title:
        role_title = _fallback_role_title(text)
    if not role_title:
        role_title = _fallback_role_title_from_caps(text)
    if not company:
        company = url_company

    skills = _match_vocabulary(text, KNOWN_SKILL_KEYWORDS)
    ats_keywords = _match_vocabulary(text, ATS_KEYWORD_VOCAB)
    experience_level = _match_first_pattern(text, EXPERIENCE_LEVEL_PATTERNS, DEFAULT_EXPERIENCE_LEVEL)
    location = _match_first_pattern(text, LOCATION_PATTERNS, DEFAULT_LOCATION)
    hiring_team_names = _extract_hiring_team_names(text)

    return ParsedJobPosting(
        roleTitle=role_title or "Not specified",
        company=company or "Not specified",
        skills=skills,
        experienceLevel=experience_level,
        location=location,
        hiringTeamNames=hiring_team_names,
        atsKeywords=ats_keywords,
        parsingConfidence="high" if role_title and company else "low",
    )
